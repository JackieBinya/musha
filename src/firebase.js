import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
import 'firebase/storage';

const firebaseConfig = firebase.initializeApp({
  apiKey: 'AIzaSyDD14Pef51sWG0czExRugPOLJO-Ne44bNI',
  authDomain: 'musha-000.firebaseapp.com',
  databaseURL: 'https://musha-000.firebaseio.com',
  projectId: 'musha-000',
  storageBucket: 'musha-000.appspot.com',
  messagingSenderId: '111803589651',
  appId: '1:111803589651:web:b90b953a233696281b4be6',
});

const storage = firebase.storage();

export { storage, firebaseConfig as firebase };
