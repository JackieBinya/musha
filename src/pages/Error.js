import React from 'react';

export const Error = () => <h2>Error</h2>;