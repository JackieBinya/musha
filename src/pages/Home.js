import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import Fuse from 'fuse.js';
import { useProperties } from '../hooks';
import { ShortProperty } from '../components/ShortProperty';
import { Hero } from '../components/Hero';
import { Loader } from '../components/Loader';
import { PropertyIconsHome } from '../components/PropertyIconsHome';

export const Home = () => {
  const { properties } = useProperties();
  const [query, setQuery] = useState('');

  const fuse = new Fuse(properties, {
    keys: ['location', 'city'],
    includeScore: true,
  });

  const results = fuse.search(query);
  const propertiesResults = query
    ? results.map((character) => character.item)
    : properties;

  return (
    <>
      <Hero query={query} setQuery={setQuery} />
      <main>
        <div className="main container">
          {propertiesResults.length === 0 ? (
            <Loader />
          ) : propertiesResults ? (
            propertiesResults.map(
              ({
                id,
                location,
                imageUrls,
                numberOfBathrooms,
                numberOfBedrooms,
                city,
                title,
                description,
              }) => (
                <li key={id} className="main-list-item">
                  <Link to={`/property/${id}`} className="main-link">
                    <ShortProperty
                      imageUrls={imageUrls}
                      title={title}
                      location={location}
                      description={description}
                    >
                      <PropertyIconsHome
                        city={city}
                        title={title}
                        location={location}
                        numberOfBathrooms={numberOfBathrooms}
                        numberOfBedrooms={numberOfBedrooms}
                      />
                    </ShortProperty>
                  </Link>
                </li>
              )
            )
          ) : (
            ''
          )}
        </div>
      </main>
    </>
  );
};
