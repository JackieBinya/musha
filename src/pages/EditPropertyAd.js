import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useForm } from 'react-hook-form';
import { firebase } from '../firebase';
import { yupResolver } from '@hookform/resolvers';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons';

import { useProperty } from '../hooks';
import { FirstSection } from '../components/FirstSection';
import { LocationSection } from '../components/LocationSection';
import { Loader } from '../components/Loader';
import { PropertyDetailsSection } from '../components/PropertyDetailsSection';
import { PROPERTY_SCHEMA } from '../constants/validations';

export const EditPropertyAds = ({ history }) => {
  const { propertyId } = useParams();
  const { property } = useProperty(propertyId);

  const { register, handleSubmit, errors } = useForm({
    resolver: yupResolver(PROPERTY_SCHEMA),
  });

  const [isTitleUpdated, setIsTitleUpdated] = useState(false);
  const [isMobileNumberUpdated, setIsMobileNumberUpdated] = useState(false);
  const [isAvailableToUpdated, setIsAvailableToUpdated] = useState(false);
  const [isCityUpdated, setIsCityUpdated] = useState(false);
  const [isLocationUpdated, setIsLocationUpdated] = useState(false);
  const [isNumberOfBathroomsUpdated, setIsNumberOfBathroomsUpdated] = useState(
    false
  );
  const [isNumberOfBedroomsUpdated, setIsNumberOfBedroomsUpdated] = useState(
    false
  );
  const [isDescriptionUpdated, setIsDescriptionUpdated] = useState(false);

  const [isModalVisible, setIsModalVisible] = useState(false);

  const onSubmit = (
    {
      city,
      location,
      availableTo,
      numberOfBathrooms,
      numberOfBedrooms,
      description,
      mobileNumber,
      title,
    },
    e
  ) => {
    const propertyRef = firebase
      .firestore()
      .collection('properties')
      .doc(propertyId);

    if (property.title !== title) {
      propertyRef
        .update({
          title,
        })
        .then(() => setIsTitleUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.availableTo !== availableTo) {
      propertyRef
        .update({
          availableTo,
        })
        .then(() => setIsAvailableToUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.mobileNumber !== mobileNumber) {
      propertyRef
        .update({
          mobileNumber,
        })
        .then(() => setIsMobileNumberUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.city !== city) {
      propertyRef
        .update({
          city,
        })
        .then(() => setIsCityUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.location !== location) {
      propertyRef
        .update({
          location,
        })
        .then(() => setIsLocationUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.description !== description) {
      propertyRef
        .update({
          description,
        })
        .then(() => setIsDescriptionUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.numberOfBedrooms !== numberOfBedrooms) {
      propertyRef
        .update({
          numberOfBedrooms,
        })
        .then(() => setIsNumberOfBedroomsUpdated(true))
        .catch((error) => console.log(error));
    }

    if (property.numberOfBathrooms !== numberOfBathrooms) {
      propertyRef
        .update({
          numberOfBathrooms,
        })
        .then(() => setIsNumberOfBathroomsUpdated(true))
        .catch((error) => console.log(error));
    }
  };

  useEffect(() => {
    if (
      isTitleUpdated ||
      isMobileNumberUpdated ||
      isAvailableToUpdated ||
      isCityUpdated ||
      isLocationUpdated ||
      isNumberOfBathroomsUpdated ||
      isNumberOfBedroomsUpdated ||
      isDescriptionUpdated
    ) {
      setIsModalVisible(true);

      setIsTitleUpdated(false);
      setIsAvailableToUpdated(false);
      setIsMobileNumberUpdated(false);
      setIsCityUpdated(false);
      setIsLocationUpdated(false);
      setIsNumberOfBathroomsUpdated(false);
      setIsNumberOfBedroomsUpdated(false);
      setIsDescriptionUpdated(false);
    }
  }, [
    isTitleUpdated,
    isAvailableToUpdated,
    isCityUpdated,
    isLocationUpdated,
    isMobileNumberUpdated,
    isDescriptionUpdated,
    isNumberOfBathroomsUpdated,
    isNumberOfBedroomsUpdated,
  ]);

  return (
    <div className="container">
      <div style={{ marginBottom: '2em' }}>
        {property && isModalVisible && (
          <div className="form-modal-container">
            <p
              className={isModalVisible ? 'show' : 'hide'}
            >{`Property ID:${propertyId} is successfully updated`}</p>
            <button
              data-testid="go-to-my-properties-action"
              type="button"
              onClick={() => history.goBack()}
            >
              {' '}
              Go to my properties
            </button>
          </div>
        )}
        <div className="property-form-container">
          {!property ? (
            <Loader />
          ) : (
            <div
              data-testid="edit-form"
              className={isModalVisible ? 'hide' : 'show'}
            >
              <button
                data-testid="back-button"
                type="button"
                onClick={() => history.goBack()}
              >
                <FontAwesomeIcon
                  icon={faChevronLeft}
                  className="property-icons-svg"
                />
                Back
              </button>

              <h1>Edit property ad</h1>

              <form onSubmit={handleSubmit(onSubmit)} className="form-primary">
                <FirstSection
                  defaultMobileNumber={property.mobileNumber}
                  defaultTitle={property.title}
                  defaultAvailableTo={property.availableTo}
                  register={register}
                  errors={errors}
                />

                <LocationSection
                  defaultCity={property.city}
                  defaultLocation={property.location}
                  register={register}
                  errors={errors}
                />

                {/* <UploadImagesSection
                  defaultImageUrls={property.imageUrls}
                  setImageUrls={setImageUrls}
                /> */}

                <PropertyDetailsSection
                  defaultNumberOfBathrooms={property.numberOfBathrooms}
                  defaultNumberOfBedrooms={property.numberOfBedrooms}
                  defaultDescription={property.description}
                  register={register}
                  errors={errors}
                />

                <button data-testid="submit-action" type="submit">
                  Save
                </button>
              </form>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};
