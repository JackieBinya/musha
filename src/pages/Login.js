import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { SignUp } from './SignUp';
import { FormMessage } from '../components/FormMessage';
import { PasswordInput } from '../components/PasswordInput';
import { EmailInput } from '../components/EmailInput';
import { firebase } from '../firebase';

export const Login = ({ history }) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [message, setMessage] = useState('');
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  const handleLogin = async (e) => {
    e.preventDefault();

    try {
      await firebase.auth().signInWithEmailAndPassword(email, password);
      setIsAuthenticated(true);
      setMessage('Congrats 🥳, you are successfully logged in.');
      setTimeout(() => history.push('/my-properties'), 5000);
    } catch (error) {
      setIsAuthenticated(false);

      if (error.code === ' auth/invalid-email') {
        setMessage('🤖 email address is not valid');
      }
      if (error.code === 'auth/user-disabled') {
        setMessage('🤖 your account has been suspended');
      }
      if (error.code === 'auth/user-not-found') {
        setMessage('🤖 email does no exist!');
      }
      if (error.code === ' auth/wrong-password') {
        setMessage('🤖 incorrect password');
      }
    }
  };
  return (
    <>
      <Link to="/signup">Sign Up</Link>
      <h1>Login</h1>
      <form className="auth-form" onSubmit={handleLogin}>
        <FormMessage
          isAuthenticated={isAuthenticated}
          message={message}
          setMessage={setMessage}
        />
        <EmailInput email={email} setEmail={setEmail} />
        <PasswordInput setPassword={setPassword} password={password} />
        <button type="submit">Continue</button>
      </form>
      <Link to="/password-reset">Have you forgotten your password?</Link>
    </>
  );
};
