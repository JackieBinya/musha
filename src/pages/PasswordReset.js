import React, { useState } from 'react';
import { firebase } from '../firebase';
import { EmailInput } from '../components/EmailInput';
import { AlertMessage } from '../components/AlertMessage';

/********************
 *  Add continue url : https://firebase.google.com/docs/auth/web/passing-state-in-email-actions#passing_statecontinue_url_in_email_actions
 */

export const PasswordReset = ({ history }) => {
  const [email, setEmail] = useState('');
  const [message, setMessage] = useState('');
  const [isEmailSent, setIsEmailSent] = useState(false);

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      await firebase.auth().sendPasswordResetEmail(email);
      setIsEmailSent(true);
      setMessage(
        'Check your inbox for further instructions on how to reset your password! '
      );
      setTimeout(() => history.push('/'), 5000);
    } catch (error) {
      setIsEmailSent(false);
      if (error.code === 'auth/invalid-email') {
        setMessage('Email invalid');
      }

      if (error.code === 'auth/user-not-found') {
        setMessage(
          'An account associated with the provided email does not exist!'
        );
      }
    }
  };

  return (
    <div className="container">
      <div className="container-inner">
        <div className="form-container">
          <AlertMessage
            isAuthenticated={isEmailSent}
            message={message}
            setMessage={setMessage}
          />
          <h1>Password Reset</h1>
          <form onSubmit={handleSubmit} className="form-primary">
            <EmailInput email={email} setEmail={setEmail} />
            <button data-testid="submit-action" type="submit">
              Reset my password
            </button>
          </form>
        </div>
      </div>
    </div>
  );
};
