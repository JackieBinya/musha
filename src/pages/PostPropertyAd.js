import React, { useState, useContext } from 'react';
import { useForm } from 'react-hook-form';
import { firebase } from '../firebase';
import { yupResolver } from '@hookform/resolvers';

import { AuthContext } from '../context/auth-context';
import { UploadImagesSection } from '../components/UploadImagesSection';
import { FirstSection } from '../components/FirstSection';
import { LocationSection } from '../components/LocationSection';
import { PropertyDetailsSection } from '../components/PropertyDetailsSection';
import { PROPERTY_SCHEMA } from '../constants/validations';
import { Footer } from '../components/Footer';

export const PostPropertyAd = ({ history }) => {
  const {
    currentUser: { uid },
  } = useContext(AuthContext);

  const { register, handleSubmit, errors } = useForm({
    resolver: yupResolver(PROPERTY_SCHEMA),
  });

  const [imageUrls, setImageUrls] = useState([]);
  const [hasSubmitted, setHasSubmitted] = useState(false);

  const onSubmit = (
    {
      city,
      location,
      availableTo,
      numberOfBathrooms,
      numberOfBedrooms,
      description,
      mobileNumber,
      title,
    },
    e
  ) => {
    firebase
      .firestore()
      .collection('properties')
      .add({
        ownerID: uid,
        city,
        location,
        availableTo,
        numberOfBathrooms,
        numberOfBedrooms,
        description,
        imageUrls,
        mobileNumber,
        title,
      })
      .then(function () {
        setHasSubmitted(true);
        // e.target.reset();
        setImageUrls([]);
      })
      .catch((error) => console.log({ error }));
  };

  return (
    <>
      <main className="container">
        <div className="container-inner">
          <div
            data-testid="alert-message"
            className={`form-modal-container ${hasSubmitted ? 'show' : 'hide'}`}
          >
            <p>Congrats! You have successfully created a property ad.</p>
            <button type="button" onClick={() => history.goBack()}>
              Go to my properties
            </button>
          </div>

          <div className={hasSubmitted ? 'hide' : 'show'}>
            <h1>Create a property ad</h1>

            <h2>Fill in the form below as accurately as possible.</h2>

            <div className="property-form-container">
              <form onSubmit={handleSubmit(onSubmit)} className="form-primary">
                <FirstSection register={register} errors={errors} />

                <LocationSection register={register} errors={errors} />

                <UploadImagesSection
                  imageUrls={imageUrls}
                  setImageUrls={setImageUrls}
                  hasSubmitted={hasSubmitted}
                />

                <PropertyDetailsSection register={register} errors={errors} />

                <button data-testid="submit-action" type="submit">
                  Submit
                </button>
              </form>
            </div>
          </div>
        </div>
      </main>
      <Footer />
    </>
  );
};
