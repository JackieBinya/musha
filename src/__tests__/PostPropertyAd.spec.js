import React from 'react';
import { firebase } from '../firebase';
import {
  render,
  cleanup,
  fireEvent,
  act,
  waitFor,
} from '@testing-library/react';
import { PostPropertyAd } from '../pages/PostPropertyAd';

import { toHaveClass } from '@testing-library/jest-dom/matchers';

expect.extend({ toHaveClass });

jest.mock('react', () => {
  const ActualReact = require.requireActual('react');
  return {
    ...ActualReact,
    useContext: () => ({
      currentUser: {
        uid: '123456',
        email: 'foo@bar.com',
      },
    }),
  };
});

jest.mock('../firebase', () => {
  return {
    firebase: {
      firestore: jest.fn(() => ({
        collection: jest.fn((properties) => ({
          add: jest.fn((data) => {
            return new Promise((resolve, reject) => {
              if (!data) reject();

              resolve();
            });
          }),
        })),
      })),
    },
  };
});

beforeEach(cleanup);

describe('<PostPropertyAd/>', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Component is successfully rendered', () => {
    const { asFragment } = render(<PostPropertyAd />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('Successful form submission', async () => {
    const { queryByLabelText, queryByTestId } = render(<PostPropertyAd />);
    const availableForRent = queryByLabelText('Rent');
    const titleInput = queryByLabelText('*Ad title:', { selector: 'input' });
    const mobileNumberInput = queryByLabelText('*Your mobile number:', {
      selector: 'input',
    });
    const cityInput = queryByLabelText('*City:', { selector: 'input' });
    const locationInput = queryByLabelText('*Location/Suburb:', {
      selector: 'input',
    });
    const selectBedrooms = queryByLabelText(
      'Choose the number of bedrooms in your property:'
    );
    const selectBathrooms = queryByLabelText(
      'Choose the number of bathrooms in your property:'
    );
    const description = queryByLabelText('*Description:', {
      selector: 'textarea',
    });
    const alertMessage = queryByTestId('alert-message');

    const submitAction = queryByTestId('submit-action');

    expect(alertMessage).toHaveClass('form-modal-container hide');

    act(() => {
      fireEvent.click(availableForRent);
      fireEvent.change(titleInput, { target: { value: 'Test Title' } });
      fireEvent.change(mobileNumberInput, { target: { value: '123456' } });
      fireEvent.change(cityInput, { target: { value: 'Test City' } });
      fireEvent.change(locationInput, { target: { value: 'Test Location' } });
      fireEvent.change(selectBedrooms, { target: { value: '1' } });
      fireEvent.change(selectBathrooms, { target: { value: '3' } });
      fireEvent.change(description, { target: { value: 'Test description' } });
    });

    fireEvent.submit(submitAction);

    // Checks that the modal becomes visible after a successful form submission
    await waitFor(() => {
      expect(alertMessage).toHaveClass('form-modal-container show');
    });
  });
});

describe('Form handles submission errors', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('AvailableTo is a required field', async () => {
    const { queryByText, queryByLabelText, queryByTestId } = render(
      <PostPropertyAd />
    );
    const availableForRent = queryByLabelText('Rent');
    const titleInput = queryByLabelText('*Ad title:', { selector: 'input' });
    const mobileNumberInput = queryByLabelText('*Your mobile number:', {
      selector: 'input',
    });
    const cityInput = queryByLabelText('*City:', { selector: 'input' });
    const locationInput = queryByLabelText('*Location/Suburb:', {
      selector: 'input',
    });
    const selectBedrooms = queryByLabelText(
      'Choose the number of bedrooms in your property:'
    );
    const selectBathrooms = queryByLabelText(
      'Choose the number of bathrooms in your property:'
    );
    const description = queryByLabelText('*Description:', {
      selector: 'textarea',
    });
    const alertMessage = queryByTestId('alert-message');

    const submitAction = queryByTestId('submit-action');

    expect(alertMessage).toHaveClass('form-modal-container hide');

    act(() => {
      fireEvent.change(titleInput, { target: { value: 'Test Title' } });
      fireEvent.change(mobileNumberInput, { target: { value: '123456' } });
      fireEvent.change(cityInput, { target: { value: 'Test City' } });
      fireEvent.change(locationInput, { target: { value: 'Test Location' } });
      fireEvent.change(selectBedrooms, { target: { value: '1' } });
      fireEvent.change(selectBathrooms, { target: { value: '3' } });
      fireEvent.change(description, { target: { value: 'Test description' } });
    });

    fireEvent.submit(submitAction);

    await waitFor(() => {
      '';
      expect(queryByText('Required!')).toBeTruthy();
    });
  });

  it('Title is a required field', async () => {
    const { queryByText, queryByLabelText, queryByTestId } = render(
      <PostPropertyAd />
    );
    const availableForRent = queryByLabelText('Rent');
    const titleInput = queryByLabelText('*Ad title:', { selector: 'input' });
    const mobileNumberInput = queryByLabelText('*Your mobile number:', {
      selector: 'input',
    });
    const cityInput = queryByLabelText('*City:', { selector: 'input' });
    const locationInput = queryByLabelText('*Location/Suburb:', {
      selector: 'input',
    });
    const selectBedrooms = queryByLabelText(
      'Choose the number of bedrooms in your property:'
    );
    const selectBathrooms = queryByLabelText(
      'Choose the number of bathrooms in your property:'
    );
    const description = queryByLabelText('*Description:', {
      selector: 'textarea',
    });
    const alertMessage = queryByTestId('alert-message');

    const submitAction = queryByTestId('submit-action');

    expect(alertMessage).toHaveClass('form-modal-container hide');

    act(() => {
      fireEvent.click(availableForRent);
      fireEvent.change(titleInput, { target: { value: null } });
      fireEvent.change(mobileNumberInput, { target: { value: '123456' } });
      fireEvent.change(cityInput, { target: { value: 'Test City' } });
      fireEvent.change(locationInput, { target: { value: 'Test Location' } });
      fireEvent.change(selectBedrooms, { target: { value: '1' } });
      fireEvent.change(selectBathrooms, { target: { value: '3' } });
      fireEvent.change(description, { target: { value: 'Test description' } });
    });

    fireEvent.submit(submitAction);

    await waitFor(() => {
      '';
      expect(queryByText('The title is required!')).toBeTruthy();
    });
  });

  it('Mobile Number is a required field', async () => {
    const { queryByText, queryByLabelText, queryByTestId } = render(
      <PostPropertyAd />
    );
    const availableForRent = queryByLabelText('Rent');
    const titleInput = queryByLabelText('*Ad title:', { selector: 'input' });
    const mobileNumberInput = queryByLabelText('*Your mobile number:', {
      selector: 'input',
    });
    const cityInput = queryByLabelText('*City:', { selector: 'input' });
    const locationInput = queryByLabelText('*Location/Suburb:', {
      selector: 'input',
    });
    const selectBedrooms = queryByLabelText(
      'Choose the number of bedrooms in your property:'
    );
    const selectBathrooms = queryByLabelText(
      'Choose the number of bathrooms in your property:'
    );
    const description = queryByLabelText('*Description:', {
      selector: 'textarea',
    });
    const alertMessage = queryByTestId('alert-message');

    const submitAction = queryByTestId('submit-action');

    expect(alertMessage).toHaveClass('form-modal-container hide');

    act(() => {
      fireEvent.click(availableForRent);
      fireEvent.change(titleInput, { target: { value: 'Test Title' } });
      fireEvent.change(mobileNumberInput, { target: { value: null } });
      fireEvent.change(cityInput, { target: { value: 'Test City' } });
      fireEvent.change(locationInput, { target: { value: 'Test Location' } });
      fireEvent.change(selectBedrooms, { target: { value: '1' } });
      fireEvent.change(selectBathrooms, { target: { value: '3' } });
      fireEvent.change(description, { target: { value: 'Test description' } });
    });

    fireEvent.submit(submitAction);

    await waitFor(() => {
      '';
      expect(queryByText('Required!')).toBeTruthy();
    });
  });

  it('City is a required field', async () => {
    const { queryByText, queryByLabelText, queryByTestId } = render(
      <PostPropertyAd />
    );
    const availableForRent = queryByLabelText('Rent');
    const titleInput = queryByLabelText('*Ad title:', { selector: 'input' });
    const mobileNumberInput = queryByLabelText('*Your mobile number:', {
      selector: 'input',
    });
    const cityInput = queryByLabelText('*City:', { selector: 'input' });
    const locationInput = queryByLabelText('*Location/Suburb:', {
      selector: 'input',
    });
    const selectBedrooms = queryByLabelText(
      'Choose the number of bedrooms in your property:'
    );
    const selectBathrooms = queryByLabelText(
      'Choose the number of bathrooms in your property:'
    );
    const description = queryByLabelText('*Description:', {
      selector: 'textarea',
    });
    const alertMessage = queryByTestId('alert-message');

    const submitAction = queryByTestId('submit-action');

    expect(alertMessage).toHaveClass('form-modal-container hide');

    act(() => {
      fireEvent.click(availableForRent);
      fireEvent.change(titleInput, { target: { value: 'Test Title' } });
      fireEvent.change(mobileNumberInput, { target: { value: '123456' } });
      fireEvent.change(cityInput, { target: { value: null } });
      fireEvent.change(locationInput, { target: { value: 'Test Location' } });
      fireEvent.change(selectBedrooms, { target: { value: '1' } });
      fireEvent.change(selectBathrooms, { target: { value: '3' } });
      fireEvent.change(description, { target: { value: 'Test description' } });
    });

    fireEvent.submit(submitAction);

    await waitFor(() => {
      expect(queryByText('The city is required!')).toBeTruthy();
    });
  });

  it('Location is a required field', async () => {
    const { queryByText, queryByLabelText, queryByTestId } = render(
      <PostPropertyAd />
    );
    const availableForRent = queryByLabelText('Rent');
    const titleInput = queryByLabelText('*Ad title:', { selector: 'input' });
    const mobileNumberInput = queryByLabelText('*Your mobile number:', {
      selector: 'input',
    });
    const cityInput = queryByLabelText('*City:', { selector: 'input' });
    const locationInput = queryByLabelText('*Location/Suburb:', {
      selector: 'input',
    });
    const selectBedrooms = queryByLabelText(
      'Choose the number of bedrooms in your property:'
    );
    const selectBathrooms = queryByLabelText(
      'Choose the number of bathrooms in your property:'
    );
    const description = queryByLabelText('*Description:', {
      selector: 'textarea',
    });
    const alertMessage = queryByTestId('alert-message');

    const submitAction = queryByTestId('submit-action');

    expect(alertMessage).toHaveClass('form-modal-container hide');

    act(() => {
      fireEvent.click(availableForRent);
      fireEvent.change(titleInput, { target: { value: 'Test Title' } });
      fireEvent.change(mobileNumberInput, { target: { value: '123456' } });
      fireEvent.change(cityInput, { target: { value: 'Test City' } });
      fireEvent.change(locationInput, { target: { value: null } });
      fireEvent.change(selectBedrooms, { target: { value: '1' } });
      fireEvent.change(selectBathrooms, { target: { value: '3' } });
      fireEvent.change(description, { target: { value: 'Test description' } });
    });

    fireEvent.submit(submitAction);

    await waitFor(() => {
      '';
      expect(queryByText('The location is required!')).toBeTruthy();
    });
  });

  it('Description is a required field', async () => {
    const { debug, queryByText, queryByLabelText, queryByTestId } = render(
      <PostPropertyAd />
    );
    const availableForRent = queryByLabelText('Rent');
    const titleInput = queryByLabelText('*Ad title:', { selector: 'input' });
    const mobileNumberInput = queryByLabelText('*Your mobile number:', {
      selector: 'input',
    });
    const cityInput = queryByLabelText('*City:', { selector: 'input' });
    const locationInput = queryByLabelText('*Location/Suburb:', {
      selector: 'input',
    });
    const selectBedrooms = queryByLabelText(
      'Choose the number of bedrooms in your property:'
    );
    const selectBathrooms = queryByLabelText(
      'Choose the number of bathrooms in your property:'
    );
    const description = queryByLabelText('*Description:', {
      selector: 'textarea',
    });
    const alertMessage = queryByTestId('alert-message');

    const submitAction = queryByTestId('submit-action');

    expect(alertMessage).toHaveClass('form-modal-container hide');

    act(() => {
      fireEvent.click(availableForRent);
      fireEvent.change(titleInput, { target: { value: 'Test Title' } });
      fireEvent.change(mobileNumberInput, { target: { value: '123456' } });
      fireEvent.change(cityInput, { target: { value: 'Test City' } });
      fireEvent.change(locationInput, { target: { value: 'Test Location' } });
      fireEvent.change(selectBedrooms, { target: { value: '1' } });
      fireEvent.change(selectBathrooms, { target: { value: '3' } });
      fireEvent.change(description, { target: { value: null } });
    });

    fireEvent.submit(submitAction);

    await waitFor(() => {
      '';
      expect(queryByText('The property description is required!')).toBeTruthy();
    });
  });
});
