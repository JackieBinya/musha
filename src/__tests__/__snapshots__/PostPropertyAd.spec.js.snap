// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostPropertyAd/> Component is successfully rendered 1`] = `
<DocumentFragment>
  <main
    class="container"
  >
    <div
      class="container-inner"
    >
      <div
        class="form-modal-container hide"
        data-testid="alert-message"
      >
        <p>
          Congrats! You have successfully created a property ad.
        </p>
        <button
          type="button"
        >
          Go to my properties
        </button>
      </div>
      <div
        class="show"
      >
        <h1>
          Create a property ad
        </h1>
        <h2>
          Fill in the form below as accurately as possible.
        </h2>
        <div
          class="property-form-container"
        >
          <form
            class="form-primary"
          >
            <section
              class="step-wrapper"
            >
              <h3
                class="step-headliner"
              >
                Miscellaneous Details
              </h3>
              <div
                class="step-contents"
              >
                <p>
                  *The property is available for :
                </p>
                <div
                  class="radio-wrapper"
                >
                  <div>
                    <label>
                      <input
                        name="availableTo"
                        type="radio"
                        value="rent"
                      />
                      Rent
                    </label>
                  </div>
                  <div>
                    <label>
                      <input
                        name="availableTo"
                        type="radio"
                        value="sale"
                      />
                      Sale
                    </label>
                  </div>
                </div>
                <div>
                  <label
                    for="title"
                  >
                    *Ad title:
                  </label>
                  <input
                    id="title"
                    name="title"
                    type="text"
                    value=""
                  />
                </div>
                <div>
                  <label
                    for="mobileNumber"
                  >
                    *Your mobile number:
                  </label>
                  <input
                    id="mobileNumber"
                    name="mobileNumber"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </section>
            <section
              class="step-wrapper"
            >
              <h3
                class="step-headliner"
              >
                Location
              </h3>
              <div
                class="step-contents"
              >
                <div>
                  <label
                    for="city"
                  >
                    *City:
                  </label>
                  <input
                    id="city"
                    name="city"
                    type="text"
                    value=""
                  />
                </div>
                <div>
                  <label
                    for="location"
                  >
                    *Location/Suburb:
                  </label>
                  <input
                    id="location"
                    name="location"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </section>
            <section
              class="step-wrapper  show "
            >
              <h3
                class="step-headliner"
              >
                Upload Images
              </h3>
              <div
                class="step-contents"
              >
                <div
                  class="preview"
                  data-testid="preview-images-wrapper"
                />
                <div>
                  <div
                    class="upload-images"
                  >
                    <div
                      class="upload-images-container show"
                      data-testid="upload-action"
                      role="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="svg-inline--fa fa-image fa-w-16 "
                        data-icon="image"
                        data-prefix="fas"
                        focusable="false"
                        role="img"
                        viewBox="0 0 512 512"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M464 448H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h416c26.51 0 48 21.49 48 48v288c0 26.51-21.49 48-48 48zM112 120c-30.928 0-56 25.072-56 56s25.072 56 56 56 56-25.072 56-56-25.072-56-56-56zM64 384h384V272l-87.515-87.515c-4.686-4.686-12.284-4.686-16.971 0L208 320l-55.515-55.515c-4.686-4.686-12.284-4.686-16.971 0L64 336v48z"
                          fill="currentColor"
                        />
                      </svg>
                    </div>
                  </div>
                  <input
                    data-testid="upload-input"
                    name="image"
                    style="display: none;"
                    type="file"
                  />
                </div>
                <p>
                  Upload up to 3 clear images of your property.
                </p>
              </div>
            </section>
            <section
              class="step-wrapper"
            >
              <h3
                class="step-headliner"
              >
                Property Details
              </h3>
              <div
                class="step-contents"
              >
                <div
                  class="step-details-wrapper"
                >
                  <label>
                    Choose the number of bedrooms in your property:
                    <select
                      name="numberOfBedrooms"
                    >
                      <option
                        value="studio"
                      >
                        Studio
                      </option>
                      <option
                        value="1"
                      >
                        1 bedroom
                      </option>
                      <option
                        value="2"
                      >
                        2 bedrooms
                      </option>
                      <option
                        value="3"
                      >
                        3+ bedrooms
                      </option>
                    </select>
                  </label>
                </div>
                <div
                  class="step-details-wrapper"
                >
                  <label>
                    Choose the number of bathrooms in your property:
                    <select
                      name="numberOfBathrooms"
                    >
                      <option
                        value="shared"
                      >
                        shared bathroom
                      </option>
                      <option
                        value="1"
                      >
                        1 bathroom
                      </option>
                      <option
                        value="2"
                      >
                        2 bathrooms
                      </option>
                      <option
                        value="3"
                      >
                        3+ bathrooms
                      </option>
                    </select>
                  </label>
                </div>
                <div
                  class="step-details-wrapper"
                >
                  <label
                    for="description"
                  >
                    *Description:
                  </label>
                  <textarea
                    id="description"
                    name="description"
                    placeholder="Type your description here..."
                  />
                </div>
              </div>
            </section>
            <button
              data-testid="submit-action"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </main>
  <footer>
    <p>
      mushaa
      <span>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-copyright fa-w-16 footer-svg"
          data-icon="copyright"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm117.134 346.753c-1.592 1.867-39.776 45.731-109.851 45.731-84.692 0-144.484-63.26-144.484-145.567 0-81.303 62.004-143.401 143.762-143.401 66.957 0 101.965 37.315 103.422 38.904a12 12 0 0 1 1.238 14.623l-22.38 34.655c-4.049 6.267-12.774 7.351-18.234 2.295-.233-.214-26.529-23.88-61.88-23.88-46.116 0-73.916 33.575-73.916 76.082 0 39.602 25.514 79.692 74.277 79.692 38.697 0 65.28-28.338 65.544-28.625 5.132-5.565 14.059-5.033 18.508 1.053l24.547 33.572a12.001 12.001 0 0 1-.553 14.866z"
            fill="currentColor"
          />
        </svg>
      </span>
      2020
    </p>
  </footer>
</DocumentFragment>
`;
